kedro/io/tensorflow_model_data_set.py

```diff
@@ -0,0 +1,237 @@
+                If prefix is not provided `file` protocol (local filesystem) will be used.
+                The prefix should be any protocol supported by ``fsspec``.
+                Note: `http(s)` doesn't support versioning.
+            load_args: TensorFlow options for loading models.
+                custom_objects: Optional dictionary mapping names (strings) to custom classes or functions to be
+                considered during deserialization.
+                compile: Boolean, whether to compile the model after loading.
+                Here you can find all available arguments:
+                https://www.tensorflow.org/api_docs/python/tf/keras/models/load_model
+                All defaults are preserved.
+            save_args: TensorFlow options for saving models.
+                Here you can find all available arguments:
+                https://www.tensorflow.org/api_docs/python/tf/keras/models/save_model
+                All defaults are preserved, except for "index", which is set to False.
+                In TF this defaults to 'tf' in TF 2.X, and 'h5' in TF 1.X.
+            version: If specified, should be an instance of
+                ``kedro.io.core.Version``. If its ``load`` attribute is
+                None, the latest version will be loaded. If its ``save``
+                attribute is None, save version will be autogenerated.
+            credentials: Credentials required to get access to the underlying filesystem.
+                E.g. for ``GCSFileSystem`` it should look like `{'token': None}`.
+            fs_args: Extra arguments to pass into underlying filesystem class.
+                E.g. for ``GCSFileSystem`` class: `{project: 'my-project', ...}`
+        """
+        _fs_args = copy.deepcopy(fs_args) or {}
+        _credentials = copy.deepcopy(credentials) or {}
+        protocol, path = get_protocol_and_path(filepath, version)
+
+        self._protocol = protocol
+        self._fs = fsspec.filesystem(self._protocol, **_credentials, **_fs_args)
+        super().__init__(
+            PurePath(filepath), version, exists_function=lambda x: Path(x).is_dir(),
+        )
+
+        self._data = None
+
+        # Handle default load and save arguments
+        self._load_args = copy.deepcopy(self.DEFAULT_LOAD_ARGS)
+        if load_args is not None:
+            self._load_args.update(load_args)
+        self._save_args = copy.deepcopy(self.DEFAULT_SAVE_ARGS)
+        if save_args is not None:
+            self._save_args.update(save_args)
+
+    def _load(self) -> None:
+        load_path = Path(self._get_load_path())
+        return tf.keras.models.load_model(str(load_path), **self._load_args)
+
+    def _save(self, data: tf.keras.Model) -> None:
+        save_path = Path(self._get_save_path())
+        tf.keras.models.save_model(data, str(save_path), **self._save_args)
+
+    def _exists(self) -> bool:
+        path = Path(self._get_load_path())
+        return path.is_dir()
```

Comment:
Why is this different to what, say, CSVDataSet does? https://github.com/quantumblacklabs/kedro/blob/develop/kedro/extras/datasets/pandas/csv_dataset.py

Will yours work with arbitrary paths like `s3://<...>`?

Reply:
it's not different, happy to take your guidance to enhance. I assume `_describe()` is for exposing useful user metadata but I don't know how its ingested elsewhere to display to user. I hadn't used kedro before the hackathon.

`_describe()` seems like nice place to expose TF summary model methods. would also be nice place to expose if its been fitted? output of summary() is a formatted chunk of text describing the model. think might just print to stdout as well. see example here https://stackoverflow.com/q/58657003

https://www.tensorflow.org/api_docs/python/tf/keras/Model#summary

will add a test case for a mocked S3. any other such arbitrary file systems that aiming to support/test?

Reply:
@mzjp2 your input about exposing the TF model summary methods here please?

Reply:
@DmitriiDeriabinQB any input on this bit about adding TF stuff to `_describe`? I think that's more about a description of the data itself, whereas `_describe` is more for _metadata_ of the dataset.

Reply:
As `fsspec` is still used here, I agree that `_describe()` should be identical to, say, `CSVDataSet`.
Also FYI, `_get_load_path()` will raise `VersionNotFoundError` if versioning is enabled and there are no versions available, hence try/except in `CSVDataSet._describe()`.

Reply:
> @DmitriiDeriabinQB any input on this bit about adding TF stuff to `_describe`? I think that's more about a description of the data itself, whereas `_describe` is more for _metadata_ of the dataset.

Is it still relevant as I don't see anything TF specific in both `_describe()` methods?

Reply:
No I haven't put anything TF specific in there yet. I wanted to ask about best place to expose before putting it into code. Seems like it would be useful information to display to the user in the interface. Although Model summary's can get arbitrarily long for each layer in model.

I am not really a Kedro user so very much keen for input about expected behaviour and functionality. Started contributing through pysprints hackathon